CMSgov/cms-carts-seds
# Call DocRaptor for PDF generation

### Description
Currently, CARTS and QMR use Prince for (some of their) PDF generation, as hosted by CMS.

This PR contains the necessary code changes for CARTS to call Prince, as hosted and managed by DocRaptor. If this PR is merged (and a cousin PR for QMR), CMS may be able to retire its Prince instance.

Outstanding considerations:
1. We would need CMS to purchase DocRaptor API keys.
   - Separate keys for CARTS and QMR, or a combined plan? If combined, we could tag documents for easy searching/sorting within DocRaptor's logs.
   - Separate keys for dev/val/prod, or combined? If combined, we could use Test Mode for the lower environments. The generated PDFs in those environments would be watermarked, which could interfere with acceptance & accessibility testing, but perhaps we could find a way to manage that as well.
   - _UPDATE: CMS purchased a single key for both apps. I added request tagging to each app. Requests from non-prod environments will always have `test: true`._
2. We would need to store those keys in AWS SSM. I have chosen the parameter name `pdf/docraptorApiKey`.
   - Within the dev AWS environment, I have saved my personal key in SSM. It is a free plan with a 5 documents/month cap, so I have hardcoded `test: true` in the request body for now.
   - Like many SSM parameters, there are 3 ways to specify this key: In your .env file (as `docraptorApiKey`), in a parameter named `/{stage}/pdf/docraptorApiKey`, or in a parameter named `/default/pdf/docraptorApiKey`. And that is the priority order. In dev, I set the default key in SSM, so it should be picked up by any PR branch.
   - _UPDATE: The CMS key is now stored in the dev environment for both CARTS and QMR, as the default SSM param. @BearHanded the key still needs to be added to val & prod._
3. DocRaptor does not have a document size limit, but it does have a 60-second time limit for synchronous PDF generation.
   - If this limit becomes a problem, we may need to look at implementing the asynchronous workflow, with its 600-second limit.
   - This will likely involve adding a Dynamo table to track request IDs, and a webhook for DocRaptor to call when generation is complete.
   - I don't think this will be a problem any time soon, but I also haven't measured. Possibly a fully-completed QMR report, printed all in a single document, would become a problem.
   - _UPDATE: It's very fast. We clocked one heavy request at under 7 seconds._
4. Since this PR does not change the underlying PDF generation service, we don't expect any regressions with regard to accessibility. Nevertheless, this must be verified before we can merge.
   - _UPDATE: We do not see any accessibility regressions._

### Related ticket(s)
CMDCT-3249

---
### How to test
1. Log into CARTS as a state user
2. Enter a report, and navigate to any section
3. Click the "Print" button at the bottom of the page. It shouldn't matter whether you print just the one section, or the entire report
4. A new tab should open, showing a watermarked PDF containing the relevant data, all pretty-like.

### Important updates
* Deprecated environment variables: `princeApiHost`, `princeApiPath`
* New environment variable: `docraptorApiKey`
* Uninstalled packages: `@smithy/signature-v4`, `@aws-crypto/sha256-js`

---
### Author checklist
